### PR TITLE
Lasers and Mirrors C++ Solution

### PR DESCRIPTION
**Pull request by maggie-j-liu**

### CODE REVIEW

File: solutions/USACO dec16-lasers-mirrors.mdx
Comment (jessechoe10) [inline solutions/USACO dec16-lasers-mirrors.mdx:25]: Totally up to you, but imo you should delete all the include headers and change it to: `#include<bits/stdc++,h>`...

File: solutions/USACO dec16-lasers-mirrors.mdx
Comment (jessechoe10) [inline solutions/USACO dec16-lasers-mirrors.mdx:10]: > The vector $\texttt{dist}$ will store the number of edges of the shortest path from the laser to each point.

"The number of edges of the shortest path"?? That doesn't make sense. Shouldn't it be "the number of edges **with** the shortest path from the laser to each point"??

Also, this is optional, but you should say the "array $\texttt{dist}$", so people who use Java won't be confused...

File: solutions/USACO dec16-lasers-mirrors.mdx
Comment (jessechoe10) [inline solutions/USACO dec16-lasers-mirrors.mdx:12]: I think you should combine the sentence " When processing an element in the queue, we determine the direction of the incoming beam." and "We will deflect this beam in the other direction, so we iterate the points that the beam can be deflected to." by saying:

"When processing an element in the queue, we will determine the direction of the incoming beam by iterating through each of the points the beam is deflecting to."

Comment (jessechoe10) [inline solutions/USACO dec16-lasers-mirrors.mdx:12]: > For each of these points, if we haven't visited it before, we push it into the queue and update the distance as one more than the current distance.

change **push** to **add**

change "if we haven't visited it before" to "if unvisited" (more concise).

File: solutions/USACO dec16-lasers-mirrors.mdx
Comment (jessechoe10) [inline solutions/USACO dec16-lasers-mirrors.mdx:14]: > If the distance to the barn, $\texttt{dist[1]}$, is $10^9$, then we aren't able to direct the beam to the barn.

If the distance to the barn, $\texttt{dist[1]}$, is $10^9$, then we print $\texttt{-1}$...

Comment (jessechoe10) [inline solutions/USACO dec16-lasers-mirrors.mdx:14]: > If we are able to, the number of mirrors needed is one less than the distance.
If the previous condition isn't true, then we print the distance to the barn - 1.

**Review by jessechoe10 — CHANGES_REQUESTED** [review]

File: solutions/USACO dec16-lasers-mirrors.mdx
Comment (jessechoe10) [inline solutions/USACO dec16-lasers-mirrors.mdx:55]: Optional: put comment line before saying `// BFS to find the minimum number of fence posts needed to direct the laser to the barn`

File: solutions/USACO dec16-lasers-mirrors.mdx
Comment (andrewwangva) [inline solutions/USACO dec16-lasers-mirrors.mdx:10]: Instead of describing your code so specifically, you can make it more concise by just saying store the direction+which fence post currently+distance.

File: solutions/USACO dec16-lasers-mirrors.mdx
Comment (andrewwangva) [inline solutions/USACO dec16-lasers-mirrors.mdx:18]: N^2???

Comment (maggie-j-liu) [reply solutions/USACO dec16-lasers-mirrors.mdx:18]: Uhh what would the time complexity be? I couldn't really figure it out; I know BFS is O(V + E), and there can possibly be an edge between any 2 points, so O(N^2)?

**Review by dongliu0426 — CHANGES_REQUESTED** [review]

File: solutions/USACO dec16-lasers-mirrors.mdx
Comment (dongliu0426) [reply solutions/USACO dec16-lasers-mirrors.mdx:18]: i'm pretty sure the tc is O(N)

File: solutions/USACO dec16-lasers-mirrors.mdx
Comment (maggie-j-liu) [reply solutions/USACO dec16-lasers-mirrors.mdx:14]: I don't use INF in my code, so it might be unclear?


I think the first one looks better?
image

File: solutions/USACO dec16-lasers-mirrors.mdx
Comment (maggie-j-liu) [reply solutions/USACO dec16-lasers-mirrors.mdx:18]: Maybe it's because I don't stop the BFS when it reaches the barn; should I add that?

File: solutions/USACO dec16-lasers-mirrors.mdx
Comment (dongliu0426) [reply solutions/USACO dec16-lasers-mirrors.mdx:14]: > I don't use INF in my code, so it might be unclear?

okay

> I think the first one looks better?

yeah, ig $\texttt{dist[1]}$ is good

File: solutions/USACO dec16-lasers-mirrors.mdx
Comment (dongliu0426) [reply solutions/USACO dec16-lasers-mirrors.mdx:18]: it's O(n) because there's a total of O(N) edges with O(N) states, so the overall tc would be O(N).

> Maybe it's because I don't stop the BFS when it reaches the barn; should I add that?

i don't think it really matters if you stop the BFS when it reaches the barn

**Review by dongliu0426 — APPROVED** [review]

**Review by jessechoe10 — APPROVED** [review]

**Comment by jessechoe10** [comment]
I think we can merge this once @andrewwangva or @dongliu0426 approves...

**Pull request merged**
